ui/page/root/home_page.go

```diff
@@ -170,6 +174,193 @@ func (hp *HomePage) OnNavigatedTo() {
 	hp.isBalanceHidden = hp.AssetsManager.IsTotalBalanceVisible()
 }
 
+type dexWalletUnlockInfo struct {
+	passEditor cryptomaterial.Editor
+	wallet     sharedW.Asset
+}
+
+// initDEX initializes a new dex client if dex is not ready.
+func (hp *HomePage) initDEX() {
+	if hp.AssetsManager.DexcReady() {
+		return // do nothing
+	}
+
+	go func() {
+		hp.AssetsManager.InitializeDEX(hp.dexCtx)
+
+		// If all went well, the dex client must be ready.
+		dexc := hp.AssetsManager.DexClient()
+		if dexc == nil {
+			return // nothing to do
+		}
+
+		// Wait until dex is ready
+		<-dexc.Ready()
+
+		loginDEXNow := dexc.Active()
+		walletsToUnlock := make(map[string]*dexWalletUnlockInfo)
+		for _, xc := range dexc.Exchanges() {
+			for _, bond := range xc.Auth.PendingBonds {
+				loginInfo, err := hp.dexWalletUnlockInfo(bond.AssetID)
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				walletsToUnlock[bond.Symbol] = loginInfo
+			}
+
+			// Always unlock current bond asset.
+			loginInfo, err := hp.dexWalletUnlockInfo(xc.Auth.BondAssetID)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			walletsToUnlock[dex.BipIDSymbol(xc.Auth.BondAssetID)] = loginInfo
+		}
+
+		if !loginDEXNow {
+			loginDEXNow = len(walletsToUnlock) > 0
+		}
+
+		if !loginDEXNow {
+			return // we are done here
+		}
+
+		dexPassEditor := hp.Theme.EditorPassword(new(widget.Editor), values.String(values.StrDexPassword))
+		dexPassEditor.Editor.SingleLine, dexPassEditor.IsRequired = true, true
+
+		// Dex has active trades, login now!
+		loginModal := modal.NewCustomModal(hp.Load).
+			Title(values.String(values.StrLogin)).
+			UseCustomWidget(func(gtx C) D {
+				extra := ""
+				if len(walletsToUnlock) > 0 {
+					extra = values.StringF(values.StrUnlockBondWalletMsg, len(walletsToUnlock))
+				}
+
+				msg := extra
+				if dexc.Active() {
+					msg = values.StringF(values.StrLoginDEXForActiveOrders, extra)
+				}
+
+				return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+					layout.Rigid(func(gtx C) D {
+						return layout.Inset{Bottom: values.MarginPadding10}.Layout(gtx, hp.Theme.Body1(msg).Layout)
+					}),
+					layout.Rigid(func(gtx C) D {
+						// Login dex editor
+						if len(walletsToUnlock) == 0 {
+							return dexPassEditor.Layout(gtx)
+						}
+
+						// Wallet login editor
+						children := []layout.FlexChild{layout.Rigid(dexPassEditor.Layout)}
+						for _, w := range walletsToUnlock {
+							children = append(children, layout.Rigid(func(gtx C) D {
+								return layout.Inset{Top: values.MarginPadding10}.Layout(gtx, w.passEditor.Layout)
+							}))
+						}
+
+						return layout.Inset{Top: values.MarginPadding5}.Layout(gtx, func(gtx C) D {
+							return layout.Flex{Axis: layout.Vertical}.Layout(gtx, children...)
```

Comment:
You don’t need to place a flex layout inside another flex layout here. The parent flex layout should suffice.

widgets := [
msg
dex password editor
]

if there are wallets to unlock, append to widgets. Then display the widgets in a single vertical flex layout.